File: Windows/HerdMe.Windows/Services/ProjectEnvironmentFile.cs
using System.Security.Cryptography;
using System.Text;

namespace HerdMe.Windows.Services;

public sealed record ProjectEnvironmentRevision(bool Exists, string? Digest)
{
    public static ProjectEnvironmentRevision Missing { get; } = new(false, null);
}

public sealed record ProjectEnvironmentDocument(
    string Contents,
    bool Exists,
    bool LoadedFromExample,
    ProjectEnvironmentRevision Revision
);

public sealed class ProjectEnvironmentChangedException : IOException
{
    public ProjectEnvironmentChangedException()
        : base("The .env file changed outside HerdMe. Reload it before saving so those changes are not overwritten.")
    {
    }
}

public static class ProjectEnvironmentFile
{
    public const int MaximumFileBytes = 4 * 1_024 * 1_024;
    private static readonly UTF8Encoding Utf8 = new(false, true);

    public static ProjectEnvironmentDocument Load(string projectPath)
    {
        var fullProjectPath = RequireProjectDirectory(projectPath);
        var environmentPath = Path.Combine(fullProjectPath, ".env");
        var examplePath = Path.Combine(fullProjectPath, ".env.example");

        if (File.Exists(environmentPath))
        {
            var data = ReadData(environmentPath);
            return new ProjectEnvironmentDocument(
                Decode(data),
                true,
                false,
                Revision(data)
            );
        }

        if (File.Exists(examplePath))
        {
            var data = ReadData(examplePath);
            return new ProjectEnvironmentDocument(
                Decode(data),
                false,
                true,
                ProjectEnvironmentRevision.Missing
            );
        }

        return new ProjectEnvironmentDocument(
            string.Empty,
            false,
            false,
            ProjectEnvironmentRevision.Missing
        );
    }

    public static ProjectEnvironmentDocument Save(
        string projectPath,
        string contents,
        ProjectEnvironmentRevision expectedRevision
    )
    {
        var fullProjectPath = RequireProjectDirectory(projectPath);
        var environmentPath = Path.Combine(fullProjectPath, ".env");
        var data = Utf8.GetBytes(contents);
        if (data.Length > MaximumFileBytes)
        {
            throw new InvalidDataException(
                "The project's .env file is larger than the supported 4 MB limit."
            );
        }

        var currentRevision = File.Exists(environmentPath)
            ? Revision(ReadData(environmentPath))
            : ProjectEnvironmentRevision.Missing;
        if (currentRevision != expectedRevision) throw new ProjectEnvironmentChangedException();

        RejectReparsePoint(environmentPath);
        var temporaryPath = Path.Combine(
            fullProjectPath,
            $".env.herdme-{Guid.NewGuid():N}.tmp"
        );
        try
        {
            using (var stream = new FileStream(
                temporaryPath,
                FileMode.CreateNew,
                FileAccess.Write,
                FileShare.None,
                64 * 1_024,
                FileOptions.WriteThrough
            ))
            {
                stream.Write(data);
                stream.Flush(true);
            }
            RejectReparsePoint(environmentPath);
            File.Move(temporaryPath, environmentPath, true);
        }
        finally
        {
            if (File.Exists(temporaryPath)) File.Delete(temporaryPath);
        }

        return new ProjectEnvironmentDocument(
            contents,
            true,
            false,
            Revision(data)
        );
    }

    private static string RequireProjectDirectory(string projectPath)
    {
        var fullProjectPath = Path.GetFullPath(projectPath);
        if (!Directory.Exists(fullProjectPath))
        {
            throw new DirectoryNotFoundException(
                "The selected project directory is no longer available."
            );
        }
        return fullProjectPath;
    }

    private static byte[] ReadData(string path)
    {
        RejectReparsePoint(path);
        var info = new FileInfo(path);
        if (!info.Exists || (info.Attributes & FileAttributes.Directory) != 0)
        {
            throw new InvalidDataException(
                "The project's .env file must be a regular UTF-8 text file."
            );
        }
        if (info.Length > MaximumFileBytes)
        {
            throw new InvalidDataException(
                "The project's .env file is larger than the supported 4 MB limit."
            );
        }
        try
        {
            return File.ReadAllBytes(path);
        }
        catch (IOException)
        {
            throw;
        }
        catch (UnauthorizedAccessException)
        {
            throw;
        }
    }

    private static string Decode(byte[] data)
    {
        try
        {
            return Utf8.GetString(data).TrimStart('\uFEFF');
        }
        catch (DecoderFallbackException error)
        {
            throw new InvalidDataException(
                "The project's .env file must be a regular UTF-8 text file.",
                error
            );
        }
    }

    private static void RejectReparsePoint(string path)
    {
        if (!File.Exists(path) && !Directory.Exists(path)) return;
        if ((File.GetAttributes(path) & FileAttributes.ReparsePoint) != 0)
        {
            throw new InvalidOperationException(
                "HerdMe will not modify a symbolic .env file. Replace it with a project-owned file first."
            );
        }
        if (Directory.Exists(path))
        {
            throw new InvalidDataException(
                "The project's .env file must be a regular UTF-8 text file."
            );
        }
    }

    private static ProjectEnvironmentRevision Revision(byte[] data)
    {
        return new ProjectEnvironmentRevision(
            true,
            Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant()
        );
    }
}

File: HerdMe/Services/ProjectEnvironmentFile.swift
import CryptoKit
import Foundation

enum ProjectEnvironmentFileError: LocalizedError, Equatable {
    case projectMissing
    case symbolicLink
    case invalidFile
    case fileTooLarge
    case changedExternally

    var errorDescription: String? {
        switch self {
        case .projectMissing:
            String(localized: "The selected project directory is no longer available.")
        case .symbolicLink:
            String(localized: "HerdMe will not modify a symbolic .env file. Replace it with a project-owned file first.")
        case .invalidFile:
            String(localized: "The project's .env file must be a regular UTF-8 text file.")
        case .fileTooLarge:
            String(localized: "The project's .env file is larger than the supported 4 MB limit.")
        case .changedExternally:
            String(localized: "The .env file changed outside HerdMe. Reload it before saving so those changes are not overwritten.")
        }
    }
}

struct ProjectEnvironmentRevision: Equatable, Sendable {
    let exists: Bool
    let digest: String?

    static let missing = ProjectEnvironmentRevision(exists: false, digest: nil)
}

struct ProjectEnvironmentDocument: Equatable, Sendable {
    let contents: String
    let exists: Bool
    let loadedFromExample: Bool
    let revision: ProjectEnvironmentRevision
}

enum ProjectEnvironmentFile {
    static let maximumFileBytes = 4 * 1_024 * 1_024

    static func load(
        projectURL: URL,
        fileManager: FileManager = .default
    ) throws -> ProjectEnvironmentDocument {
        try requireProjectDirectory(projectURL, fileManager: fileManager)
        let environmentURL = projectURL.appendingPathComponent(".env", isDirectory: false)
        let exampleURL = projectURL.appendingPathComponent(".env.example", isDirectory: false)

        if fileManager.fileExists(atPath: environmentURL.path) {
            let data = try readData(environmentURL)
            return ProjectEnvironmentDocument(
                contents: try decode(data),
                exists: true,
                loadedFromExample: false,
                revision: revision(for: data)
            )
        }

        if fileManager.fileExists(atPath: exampleURL.path) {
            let data = try readData(exampleURL)
            return ProjectEnvironmentDocument(
                contents: try decode(data),
                exists: false,
                loadedFromExample: true,
                revision: .missing
            )
        }

        return ProjectEnvironmentDocument(
            contents: "",
            exists: false,
            loadedFromExample: false,
            revision: .missing
        )
    }

    static func save(
        _ contents: String,
        projectURL: URL,
        expectedRevision: ProjectEnvironmentRevision,
        fileManager: FileManager = .default
    ) throws -> ProjectEnvironmentDocument {
        try requireProjectDirectory(projectURL, fileManager: fileManager)
        let environmentURL = projectURL.appendingPathComponent(".env", isDirectory: false)
        let data = Data(contents.utf8)
        guard data.count <= maximumFileBytes else {
            throw ProjectEnvironmentFileError.fileTooLarge
        }

        let currentRevision: ProjectEnvironmentRevision
        if fileManager.fileExists(atPath: environmentURL.path) {
            let currentData = try readData(environmentURL)
            currentRevision = revision(for: currentData)
        } else {
            currentRevision = .missing
        }
        guard currentRevision == expectedRevision else {
            throw ProjectEnvironmentFileError.changedExternally
        }

        try rejectSymbolicLink(environmentURL)
        let temporaryURL = projectURL.appendingPathComponent(
            ".env.herdme-\(UUID().uuidString.lowercased()).tmp",
            isDirectory: false
        )
        defer { try? fileManager.removeItem(at: temporaryURL) }
        try data.write(to: temporaryURL, options: [.atomic])
        try fileManager.setAttributes(
            [.posixPermissions: NSNumber(value: Int16(0o600))],
            ofItemAtPath: temporaryURL.path
        )
        if fileManager.fileExists(atPath: environmentURL.path) {
            _ = try fileManager.replaceItemAt(
                environmentURL,
                withItemAt: temporaryURL,
                backupItemName: nil,
                options: []
            )
        } else {
            try fileManager.moveItem(at: temporaryURL, to: environmentURL)
        }

        return ProjectEnvironmentDocument(
            contents: contents,
            exists: true,
            loadedFromExample: false,
            revision: revision(for: data)
        )
    }

    private static func requireProjectDirectory(
        _ projectURL: URL,
        fileManager: FileManager
    ) throws {
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: projectURL.path, isDirectory: &isDirectory),
            isDirectory.boolValue
        else {
            throw ProjectEnvironmentFileError.projectMissing
        }
    }

    private static func readData(_ url: URL) throws -> Data {
        try rejectSymbolicLink(url)
        guard let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey]),
            values.isRegularFile == true
        else {
            throw ProjectEnvironmentFileError.invalidFile
        }
        guard (values.fileSize ?? 0) <= maximumFileBytes else {
            throw ProjectEnvironmentFileError.fileTooLarge
        }
        do {
            return try Data(contentsOf: url, options: [.mappedIfSafe])
        } catch {
            throw ProjectEnvironmentFileError.invalidFile
        }
    }

    private static func decode(_ data: Data) throws -> String {
        guard data.count <= maximumFileBytes,
            let contents = String(data: data, encoding: .utf8)
        else {
            throw data.count > maximumFileBytes
                ? ProjectEnvironmentFileError.fileTooLarge
                : ProjectEnvironmentFileError.invalidFile
        }
        return contents.hasPrefix("\u{FEFF}") ? String(contents.dropFirst()) : contents
    }

    private static func rejectSymbolicLink(_ url: URL) throws {
        if (try? url.resourceValues(forKeys: [.isSymbolicLinkKey]).isSymbolicLink) == true {
            throw ProjectEnvironmentFileError.symbolicLink
        }
    }

    private static func revision(for data: Data) -> ProjectEnvironmentRevision {
        ProjectEnvironmentRevision(
            exists: true,
            digest: SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        )
    }
}
